stdlib/src/builtin/string_literal.mojo

```diff
@@ -356,23 +361,43 @@ struct StringLiteral(
         """
         return self.__str__()
 
-    fn __iter__(ref [_]self) -> _StringSliceIter[StaticConstantOrigin]:
-        """Return an iterator over the string literal.
+    fn __iter__[
+        is_mutable: Bool = False,
+        origin: Origin[is_mutable]
+        .type = _lit_mut_cast[StaticConstantOrigin, is_mutable]
+        .result,
+    ](self) -> _StringSliceIter[origin]:
+        """Iterate over the string unicode characters.
+
+        Parameters:
+            is_mutable: Whether the result will be mutable.
+            origin: The origin of the data.
 
         Returns:
-            An iterator over the string.
+            An iterator of references to the string unicode characters.
         """
-        return _StringSliceIter[StaticConstantOrigin](
+        constrained[not is_mutable, "StringLiteral can't be mutated"]()
+        return _StringSliceIter[origin](
             unsafe_pointer=self.unsafe_ptr(), length=self.byte_length()
         )
 
-    fn __reversed__(self) -> _StringSliceIter[StaticConstantOrigin, False]:
-        """Iterate backwards over the string, returning immutable references.
+    fn __reversed__[
+        is_mutable: Bool = False,
+        origin: Origin[is_mutable]
+        .type = _lit_mut_cast[StaticConstantOrigin, is_mutable]
+        .result,
+    ](self) -> _StringSliceIter[origin, forward=False]:
+        """Iterate backwards over the string unicode characters.
+
+        Parameters:
+            is_mutable: Whether the result will be mutable.
+            origin: The origin of the data.
 
         Returns:
-            A reversed iterator over the string.
+            A reversed iterator of references to the string unicode characters.
         """
-        return _StringSliceIter[StaticConstantOrigin, False](
+        constrained[not is_mutable, "StringLiteral can't be mutated"]()
```

Comment:
I don't see how we'd do that given that the origin parameter in `UnsafePointer` is declared `origin: Origin[True].type = MutableAnyOrigin`. I guess We'd need to add `is_mutable = True, origin: Origin[is_mutable].type = _lit_mut_cast[MutableAnyOrigin, is_mutable].result`. That way it would either be `MutableAnyOrigin` or `ImmutableAnyOrigin` and once you bitcast the pointer to `StaticConstantOrigin` it wouldn't allow mutation (?).

I'd like to look into it in another PR but how would I test that it's carrying the correct origin?